auth/msal.py

```diff
@@ -0,0 +1,80 @@
+import requests
+from flask import session, request, redirect, url_for
+import msal
+import config
+
+
+def login():
+    # Technically we could use empty list [] as scopes to do just sign in,
+    # here we choose to also collect end user consent upfront
+    session["flow"] = build_auth_code_flow(scopes=config.SCOPE)
+    return redirect(session["flow"]["auth_uri"]), 302
+
+
+def logout():
+    post_logout_redirect_uri = request.args.get("post_logout_redirect_uri", "")
+    session.clear()  # Wipe out user and its token cache from session
+    return redirect(  # Also logout from your tenant's web session
+        config.AUTHORITY + "/oauth2/v2.0/logout" +
+        "?post_logout_redirect_uri=" + post_logout_redirect_uri)
+
+
+# The absolute URL that points here must match your app's redirect_uri set in AAD
+def get_token():
+    try:
+        cache = _load_cache()
+        result = _build_msal_app(cache=cache).acquire_token_by_auth_code_flow(
+            session.get("flow", {}), request.args)
+        if "error" in result:
+            return result, 500
+        session["user"] = result.get("id_token_claims")
+        _save_cache(cache)
+        return session["user"], 200
+    except ValueError:  # Usually caused by CSRF
+        pass  # Simply ignore them
```

Comment:
:thinking: Might even be worth using the warning module so that the person deploying this repo is aware of this problem...